core/services/relay/evm/mercury/wsrpc/cache/cache.go

```diff
@@ -0,0 +1,397 @@
+	v.Lock()
+	defer v.Unlock()
+	v.val = val
+	v.err = err
+	v.expiresAt = expiresAt
+	close(v.fetchCh)
+	v.fetchCh = nil
+	v.fetching = false
+}
+
+func (v *cacheVal) abandonFetch(err error) {
+	v.completeFetch(nil, err, time.Now())
+}
+
+func (v *cacheVal) waitForResult(ctx context.Context, chResult <-chan struct{}, chStop <-chan struct{}) (*pb.LatestReportResponse, error) {
+	select {
+	case <-ctx.Done():
+		_, err := v.read()
+		return nil, errors.Join(err, ctx.Err())
+	case <-chStop:
+		return nil, ErrStopped{}
+	case <-chResult:
+		return v.read()
+	}
+}
+
+// memCache stores values in memory
+// it will never return a stale value older than latestPriceTTL, instead
+// waiting for a successful fetch or caller context cancels, whichever comes
+// first
+type memCache struct {
+	services.StateMachine
+	lggr logger.Logger
+
+	client Client
+
+	latestPriceTTL       time.Duration
+	maxStaleAge          time.Duration
+	latestReportDeadline time.Duration
+
+	cache sync.Map
+
+	wg     sync.WaitGroup
+	chStop services.StopChan
+}
+
+func newMemCache(client Client, cfg Config) *memCache {
+	return &memCache{
+		services.StateMachine{},
+		cfg.Logger.Named("MercuryMemCache"),
+		client,
+		cfg.LatestReportTTL,
+		cfg.MaxStaleAge,
+		cfg.LatestReportDeadline,
+		sync.Map{},
+		sync.WaitGroup{},
+		make(chan (struct{})),
+	}
+}
+
+type ErrStopped struct{}
+
+func (e ErrStopped) Error() string {
+	return "memCache was stopped"
+}
+
+// LatestReport
+// NOTE: This will actually block on all types of errors, even non-timeouts.
+// Context should be set carefully and timed to be the maximum time we are
+// willing to wait for a result, the background thread will keep re-querying
+// until it gets one even on networking errors etc.
+func (m *memCache) LatestReport(ctx context.Context, req *pb.LatestReportRequest) (resp *pb.LatestReportResponse, err error) {
+	if req == nil {
+		return nil, errors.New("req must not be nil")
+	}
+	if m.latestPriceTTL <= 0 {
+		return m.client.RawClient().LatestReport(ctx, req)
+	}
+	vi, _ := m.cache.LoadOrStore(req, &cacheVal{
+		sync.RWMutex{},
+		false,
+		nil,
+		nil,
+		nil,
+		time.Now(), // first result is always "expired" and requires fetch
+	})
+	v := vi.(*cacheVal)
+
+	// HOT PATH
+	v.RLock()
+	if time.Now().Before(v.expiresAt) {
+		// CACHE HIT
+		promCacheHitCount.WithLabelValues(m.client.ServerURL(), mercuryutils.BytesToFeedID(req.FeedId).String()).Inc()
+
+		defer v.RUnlock()
+		return v.val, nil
+	} else if v.fetching {
+		// CACHE WAIT
+		promCacheWaitCount.WithLabelValues(m.client.ServerURL(), mercuryutils.BytesToFeedID(req.FeedId).String()).Inc()
+		// if someone else is fetching then wait for the fetch to complete
+		ch := v.fetchCh
+		v.RUnlock()
+		return v.waitForResult(ctx, ch, m.chStop)
+	} else {
+		// CACHE MISS
+		promCacheMissCount.WithLabelValues(m.client.ServerURL(), mercuryutils.BytesToFeedID(req.FeedId).String()).Inc()
+		// fallthrough to cold path and fetch
+		v.RUnlock()
```

Comment:
no need to unclock here and open for a concurrent goroutine to acquire the value lock.